Reworked the "Headless Core" section of the ADVANCED guide (#227)
elastic/search-ui: ADVANCED.md

- [Headless Core Reference](#headless-core-reference)

## Headless Core Concepts

```
                                    |
    @elastic/react-search-ui        |   @elastic/search-ui
                                    |
                                    |
          SearchProvider <--------------- SearchDriver
              |     |               |          |
   State /    |     |               |          | State /
   Actions    |     |               |          | Actions
              |     |               |          |
        Components  |               |          |
              |     |               |          |
              v     v               |          v
------------------------------------+----------------------------
              |     |                          |
              v     v                          v
          Using     Headless Usage       Headless Usage outside
     Components     in React             of React
```

The core is a separate, vanilla JS library which can be used for any JavaScript based implementation.

> [@elastic/search-ui](https://github.com/elastic/search-ui/tree/master/packages/search-ui)

> One way to think of the Headless Core is as a search experience without a view. Instead of providing UI like search boxes

Reviewer: Here's how I'd write this, maybe @goodroot can find a compromise

> The Headless Core implements the functionality behind a search experience, but without its own view. It provides the underlying "state" and "actions" associated with that UI. For instance, the core provides a `setSearchTerm` action, which can be used to save a `searchTerm` property in the state. Calling the action using the value of an `<input>` will save the `searchTerm` to be used to build a query.

Author: Used it with just a few small tweaks.

and dropdowns, it provides the underlying "state" and "actions" associated with that UI. For instance, instead of
a "search box", the core has a `searchTerm` property which represents the value of the search box (the state), and a
`setSearchTerm` method to update that searchTerm (the action).

> Search UI lets you work directly with "state" and "actions", which means you are not restricted to using

Reviewer: All of the components in this library use the Headless Core under the hood. For instance, Search UI provides `SearchBox` component for collecting input from a user. But you are not restricted to using just that component. Since Search UI lets you work directly with "state" and "actions", you could use any type of input you want! As long as your input or component calls the Headless Core's `setSearchTerm` action, it will "just work". This gives you maximum flexibility over your experience if you need more than the components in Search UI have to offer.

Author: Used it with just a few small tweaks.

JUST a `SearchBox` component for collecting input from a user. You could use any type of input you want! As long
as it ends up calling `setSearchTerm`, it will "just work". This gives you maximum flexibility over your experience.

> All of the components in this library use the headless core under the hood. You too can work directly with the headless

Reviewer: I moved this up 🤷‍♂

Author: 👌

core if you need more than the components offer.

> The `SearchProvider` is just a React wrapper around the Headless Core, and makes state and actions available to Search UI

Reviewer: ```suggestion
The `SearchProvider` is a React wrapper around the Headless Core, and makes state and actions available to Search UI
```

</SearchProvider>
```

## Working with the Headless Core

If you wish to work with Search UI outside of a particular Component, you'll work
directly with the core.

There's 3 ways you can do this:

### 1. The Render Prop on SearchProvider

When you configure `SearchProvider`, you need to provide a function as the child of the provider.

That function is actually a [Render Prop](https://reactjs.org/docs/render-props.html) that exposes the Context for you to work with.

> On use case for that would be to render a "loading" indicator any time the application is fetching data.

Reviewer 2: ```suggestion
One use case for that would be to render a "loading" indicator any time the application is fetching data.
```

</SearchProvider>
```

## Working with the Headless Core

If you wish to work with Search UI outside of a particular Component, you'll work
directly with the core.

There's 3 ways you can do this:

### 1. The Render Prop on SearchProvider

When you configure `SearchProvider`, you need to provide a function as the child of the provider.

That function is actually a [Render Prop](https://reactjs.org/docs/render-props.html) that exposes the Context for you to work with.

> A good use case for that could be to render a "loading" indicator any time the application is fetching data.

Reviewer: ```suggestion
One use case for that would be to render a "loading" indicator any time the application is fetching data.
```

Author: Used it.

### State

State can be divided up into a few different types.

1. Request State - State that is used as parameters on Search API calls.
2. Result State - State that represents a response from a Search API call.
3. Application State - The general state.

> Request state and Result State will often have similar values. For instance, `searchTerm` and `resultSearchTerm`.

Reviewer 2: ```suggestion
Request State and Result State will often have similar values. For instance, `searchTerm` and `resultSearchTerm`.
```

Shows a dropdown for selecting the number of results to show per page.

> Uses [20, 40, 60] as a default options. You can use `options` prop to pass custom options.

Reviewer 2: ```suggestion
Uses [20, 40, 60] as default options. You can use `options` prop to pass custom options.
```
